OCPBUGS-44199: Allow spaces in AWS tags

patrickdillon opened the pull request:
An incorrect regex validation prevents users from specifying AWS tag keys or values that include spaces, which are
allowed by AWS's official regex:

https://docs.aws.amazon.com/directoryservice/latest/devguide/API_Tag.html

This bug is affecting users, so it would be good to backport to at least 4.17.

patrickdillon commented:
/jira refresh

barbacbd reviewed (commented):
/lgtm

everettraven commented on config/v1/types_infrastructure.go:
```diff
@@ -531,15 +531,15 @@ type AWSResourceTag struct {
 	// +kubebuilder:validation:MinLength=1
 	// +kubebuilder:validation:MaxLength=128
-	// +kubebuilder:validation:Pattern=`^[0-9A-Za-z_.:/=+-@]+$`
+	// +kubebuilder:validation:Pattern=`^([\p{L}\p{Z}\p{N}_.:/=+\-@]*)$`
```
I know that this is updating the regex to match the pattern in the AWS tag documentation, but have some thoughts/questions:

- The use of the unicode category syntax format is not one that I've seen commonly used throughout our APIs, and I imagine is likely to be unfamiliar to those that see this in the future. I only know what this format is after googling it and finding https://www.regular-expressions.info/unicode.html#prop that has a section explaining it. Would it be possible to achieve the same validation using more commonly known values in the regular expression?
- Conventions have changed a bit and we encourage the use of CEL expressions, via the `kubebuilder:validation:XValidation` marker, for pattern matching so that we can provide users a more clear error message as opposed to returning the pattern to the user. Could you update this to use a CEL expression for the pattern validation with a helpful error message for the user when the pattern doesn't match?

everettraven replied:
Another thing I think might be important to consider since this is a more uncommon regex pattern - if we don't change this, are there any tests that can verify that it is working as expected?

From what I can tell, the pattern in the AWS docs is taken from their Java pattern. Java seems to use a unicode-based regex engine, and Go uses RE2. While Go may support this unicode category value in a regex string, I'm not sure if it will function the same way it would in Java.

patrickdillon replied:
@everettraven Thanks for the review. I updated everything as you suggested, can you PTAL? 

I also add the validation that keys may not be prefixed with `aws:`

everettraven commented on config/v1/types_infrastructure.go:
```diff
@@ -531,15 +531,15 @@ type AWSResourceTag struct {
 	// +kubebuilder:validation:MinLength=1
 	// +kubebuilder:validation:MaxLength=256
-	// +kubebuilder:validation:Pattern=`^[0-9A-Za-z_.:/=+-@]+$`
+	// +kubebuilder:validation:Pattern=`^([\p{L}\p{Z}\p{N}_.:/=+\-@]*)$`
```
Same comments regarding the previous pattern apply here

everettraven commented on config/v1/types_infrastructure.go:
```diff
@@ -531,15 +531,15 @@ type AWSResourceTag struct {
 	// key is the key of the tag
```
Looks like the documentation of this field is a bit sparse. While you are making changes here, it would be good to update the godoc here to explain a bit more to a user how this field is used and what the constraints are. Users aren't able to see the same markers we are, so a good description here goes a long way in making the user experience of using `oc/kubectl explain` to learn more about the APIs they are interacting with better.

A good guideline for some things to consider when writing godoc for a field can be found here: https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#write-user-readable-documentation-in-godoc

everettraven commented on config/v1/types_infrastructure.go:
```diff
@@ -531,15 +531,15 @@ type AWSResourceTag struct {
+	// +kubebuilder:validation:Pattern=`^([\p{L}\p{Z}\p{N}_.:/=+\-@]*)$`
 	// +required
 	Key string `json:"key"`
 	// value is the value of the tag.
 	// Some AWS service do not support empty values. Since tags are added to resources in many services, the
 	// length of the tag value must meet the requirements of all services.
```
Ditto re: godoc updates

everettraven commented on config/v1/types_infrastructure.go:
```diff
@@ -528,18 +528,21 @@ type AWSPlatformStatus struct {
 	// +kubebuilder:validation:MinLength=1
 	// +kubebuilder:validation:MaxLength=128
-	// +kubebuilder:validation:Pattern=`^[0-9A-Za-z_.:/=+-@]+$`
+	// +kubebuilder:validation:XValidation:rule=`self.matches('^(?!aws:).+[0-9A-Za-z_.:/=+-@ ]+$')`,message="invalid AWS resource tag key. The string can contain only the set of Unicode letters, digits, white-space, '_', '.', '/', '=', '+', '-', ':', '@' and cannot be prefixed with 'aws:'"
```
Why can't this be prefixed with `aws:` ? This doesn't seem to have been prevented before. What is the potential of existing resources having a key value with this prefix, making it invalid with the new validation?

patrickdillon replied:
> Why can't this be prefixed with aws: ? This doesn't seem to have been prevented before. 

I noticed it when reading the docs https://docs.aws.amazon.com/directoryservice/latest/devguide/API_Tag.html
Apparently the RE2 regexes do not support this lookahead which is why the test is failing (although I thought I tested this locally before pushing!), so writing such a regex is quite tedious and unreadable. So would you be fine with just dropping this?

Users are not allowed to input these values in the API, it can only be set by the installer, so we can easily add validation in the installer to check for this prefix. WDYT?

patrickdillon replied:
I pushed an updated version without this check. Like I said, I think it makes the most sense to add this check in the installer.

everettraven replied:
I think it would make sense to add the prefix check if it is already not allowed. 

I'm new to doing API reviews for OpenShift, but this seems like a case where ratcheting _could_ make sense if we are confident that doing so wouldn't negatively impact users. Your controller would still have to support any handling of previously supported values though (i.e if your controller returns an error in the status if prefixed with `aws:` you can't remove that logic just because you ratcheted the validation).

To really know if this is "safe" to ratchet we probably need to gather a bit more insights to see if there are any instances of this resource on a cluster where this value is prefixed with `aws:`, and if so how many.

For reference, you can also utilize available CEL functions to check if a string starts with a particular string. See: https://github.com/google/cel-spec/blob/master/doc/langdef.md#string-functions (there is a `startsWith` in that section)

You can also chain the `kubebuilder:validation:XValidation` markers to have multiple validation rules that are ANDed together

everettraven replied:
Something I realize I forgot to ask in ^ - what happens today if the prefix is included?

patrickdillon replied:
> Something I realize I forgot to ask in ^ - what happens today if the prefix is included?

I just tested and install fails immediately due to a failure from the aws sdk:

> ERROR failed to fetch Cluster: failed to generate asset "Cluster": failed to create cluster: failed during pre-provisioning: failed to create IAM roles: failed to create IAM master role: failed to create master role: InvalidInput: Tag keys beginning with aws: are reserved for system use.

Tags can only be set in the install config--they can't be changed day 2. This definitely makes sense to validate in the installer. 

if we add the prefix check here, it would simply be future proofing for a feature where we extend the API to make mutable day-2 tags, which I don't think is on the horizon. I'm totally fine adding the prefix check, particularly with your tip for CEL string functions, but it has no real benefit.

JoelSpeed replied:
Even though we don't currently support updating this on day 2, it can still be modified by users, `oc edit --subresource status` would allow them to manipulate the values, so there is benefit in having that validation here as well

everettraven commented on config/v1/types_infrastructure.go:
```diff
@@ -528,18 +528,21 @@ type AWSPlatformStatus struct {
 	// +kubebuilder:validation:MinLength=1
 	// +kubebuilder:validation:MaxLength=128
-	// +kubebuilder:validation:Pattern=`^[0-9A-Za-z_.:/=+-@]+$`
+	// +kubebuilder:validation:XValidation:rule=`self.matches('^(?!aws:).+[0-9A-Za-z_.:/=+-@ ]+$')`,message="invalid AWS resource tag key. The string can contain only the set of Unicode letters, digits, white-space, '_', '.', '/', '=', '+', '-', ':', '@' and cannot be prefixed with 'aws:'"
```
Additionally, could you add tests for the validations to make sure:
- Values we expect to be valid are actually valid
- Values we expect to be invalid are not valid

patrickdillon replied:
https://github.com/openshift/api/tree/master/tests does not tell you how to actually execute the tests

patrickdillon replied:
ah I see that there is a nested hack dir: https://github.com/openshift/api/blob/master/tests/hack/test.sh but that is panicking because I don't have the envtest dependencies

patrickdillon replied:
ah, so I just needed to use `make` and that runs all the tests. that pulled in my new tests which failed (as expected), so that's good. it took 16m, so unless there is some caching that takes place on subsequent runs I need to figure out how to run individual tests

everettraven replied:
I'm newer to the API review process for OpenShift so I don't know off the top of my head how to run individual tests. Maybe @JoelSpeed would have something handy?

JoelSpeed replied:
So I normally use the Makefile in each API group, eg https://github.com/openshift/api/blob/master/config/v1/Makefile

With `make -C config/v1 test`, or you can do what it is doing yourself if you want

```
GINKGO_EXTRA_ARGS="--focus=something that is in my test case" make -C tests test
```

TBH for dev I normally just adjust the focus within the makefile myself

everettraven commented on config/v1/types_infrastructure.go:
```diff
@@ -528,18 +528,21 @@ type AWSPlatformStatus struct {
-	// key is the key of the tag
+	// key sets the key of the AWS resource tag key-value pair. Key is required when defining an AWS resource tag.
+	// Length Constraints: Minimum length of 1. Maximum length of 128.
+	// The string can contain only the set of Unicode letters, digits, white-space, '_', '.', '/', '=', '+', '-', ':', '@' and cannot be prefixed with 'aws:'.
```
The validation pattern no longer looks like it technically supports all unicode letters, digits, and whitespace. Maybe something like:
```suggestion
	// The string can contain only the set of alphanumeric characters, space (' '), '_', '.', '/', '=', '+', '-', ':', '@' and cannot be prefixed with 'aws:'.
```
?

This will likely also need to be updated further depending on any changes related to the prefix ratcheting.

patrickdillon commented:
@everettraven @JoelSpeed I think the API changes are complete, but I am blocked trying to add tests. I have added a commit with some of the tests I have been experimenting with, but here is what I have found:

1. It seems that specifying a featuregate is required. If I don't specify a featuregate, the tests fail:

```
the following fields were expected to match but did not:
  [(status.platformStatus.aws.cloudLoadBalancerConfig/status.platformStatus.aws.cloudLoadBalancerConfig)]
  Expected
      <*unstructured.Unstructured | 0xc00008c800>: {
          Object: {
....
```

Not sure what I should do here, as there is no associated feature gate. Right now I've added an arbitrary feature gate to get the one test I can write to pass. This seems like a bug.

2. Another issue is that when I test the `aws:`- prefix validation (expected error) with `onCreate`, I get the same error. So next I try testing with `onUpdate`. If I put the `aws:key` key in the `initial` config, the test fails:

```
  [38;5;9m[FAILED] initial object status should update successfully
   ...
              Message: "Infrastructure.config.openshift.io \"test-xw97d\" is invalid: status.platformStatus.aws.resourceTags[0].key: Invalid value: \"string\": the prefix 'aws:' is reserved for AWS system usage and cannot be used at the beginning of a key",
``` 

This error message actually demonstrates the validation seems to be working., I just can't figure out a test to confirm it.  When I try to update the test so that the bad key is only used during `updated," no error is returned. I wonder if this has something to do with how for this object, the values in question are in the `status` and not the `spec`?

Please LMK what you think and next steps.

JoelSpeed commented:
> It seems that specifying a featuregate is required. If I don't specify a featuregate, the tests fail:

Can you push the tests as you think they should look, and we can verify the formatting. Typically, tests that aren't gated, should be run from the AAA_ungated test file. Otherwise they should be in a file named after the gate.

> Another issue is that when I test the aws:- prefix validation (expected error) with onCreate, I get the same error. So next I try testing with onUpdate. If I put the aws:key key in the initial config, the test fails:

For status tests, we typically use `onUpdate` tests, have a look at the readme in the tests dir and it explains this.
You create an `initial` state, with some base status`, then `updated` is used to add the field for the validation you want to test, and the you can `expectedStatusError` to check against an error in updating the status subresource

patrickdillon commented:
Thanks Joel! You cleared up both of my issues. As you suggested, the problem for the second part was that I was using `expectedError` instead of `expectedStatusError` :facepalm: 

All feedback has been addressed so this should be ready for another look.

everettraven commented on config/v1/tests/infrastructures.config.openshift.io/AAA_ungated.yaml:
```diff
@@ -1365,3 +1365,177 @@ tests:
+              region: us-east-1
+              resourceTags:
+              - key: key:_./=+-@
+                value: value with space
```
Should the value field validation for these same characters be flexed in this test case as well?

everettraven commented on config/v1/tests/infrastructures.config.openshift.io/AAA_ungated.yaml:
```diff
@@ -1365,3 +1365,177 @@ tests:
+              region: us-east-1
+              resourceTags:
+              - key: key:_./=+-@*
+                value: value with space
```
Add a separate test for value field with invalid value?

everettraven commented on config/v1/types_infrastructure.go:
```diff
@@ -528,18 +528,22 @@ type AWSPlatformStatus struct {
 type AWSResourceTag struct {
-	// key is the key of the tag
+	// key sets the key of the AWS resource tag key-value pair. Key is required when defining an AWS resource tag.
+	// Length Constraints: Minimum length of 1. Maximum length of 128.
```
Nit:
```suggestion
	// Minimum length of 1. Maximum length of 128.
```

everettraven commented on config/v1/types_infrastructure.go:
```diff
@@ -528,18 +528,22 @@ type AWSPlatformStatus struct {
 	// +required
 	Key string `json:"key"`
-	// value is the value of the tag.
+	// value sets the value of the AWS resource tag key-value pair. Value is required when defining an AWS resource tag.
+	// The string can contain only the set of Unicode letters, digits, white-space, '_', '.', '/', '=', '+', '-', ':','@'.
```
Missing length constraint information in the documentation here.

everettraven reviewed (commented):
A few more things. Other than this, I think the changes look good.

patrickdillon commented on config/v1/tests/infrastructures.config.openshift.io/AAA_ungated.yaml:
```diff
@@ -1365,3 +1365,191 @@ tests:
+          cpuPartitioning: None
+          infrastructureTopology: HighlyAvailable
+          platform: AWS
+          platformStatus:
+            aws:
```
I'm not sure what to do here with the `cloudLoadBalancerConfig`. The tests seem to be failing both when I include it and when I don't. I am not including it, because the field is feature gated 

https://github.com/patrickdillon/api/blob/f4e9b1854e85290d8adb4ae4494496e1a639eb33/config/v1/types_infrastructure.go#L523

And that feature gate is not in the default feature set, so the field shouldn't be included here, right? 

But the failing test complains about it:

```
[0m [38;5;243m[config.openshift.io/v1, Resource=infrastructures][ClusterProfiles=Hypershift,SelfManagedHA][FeatureSet="DevPreviewNoUpgrade"][FeatureGate=][File=0000_10_config-operator_01_infrastructures-DevPreviewNoUpgrade.crd.yaml] Infrastructure [0mOn Update [38;5;9m[1m[It] Should be able to create an aws resourcetag with characters '_', '.', '/', '=', '+', '-', ':', '@'[0m
[38;5;243m/home/padillon/go/src/github.com/openshift/api/tests/generator.go:314[0m

  [38;5;9m[FAILED] the following fields were expected to match but did not:
  [(status.platformStatus.aws.cloudLoadBalancerConfig/status.platformStatus.aws.cloudLoadBalancerConfig)]
```

So I try adding it like below, and that also fails...

```yaml
      expected: |
        apiVersion: config.openshift.io/v1
        kind: Infrastructure
        spec:
          platformSpec:
            type: AWS
            aws: {}
        status:
          controlPlaneTopology: HighlyAvailable
          cpuPartitioning: None
          infrastructureTopology: HighlyAvailable
          platform: AWS
          platformStatus:
            aws:
              cloudLoadBalancerConfig:
                dnsType: PlatformDefault
              region: us-east-1
              resourceTags:
              - key: key with space
                value: value with space
            type: AWS
```

patrickdillon replied:
@JoelSpeed it seems that the problem is that tests for TechPreview (etc) as well as the Default feature set are run on the AAA_ungated tests. So if I don't include the `cloudLoadBalancerConfig` [the TechPreview (etc) tests fail](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_api/2124/pull-ci-openshift-api-master-integration/1881465800062144512), but if I do include the `cloudLoadBalancerConfig` tests for the Default set fail (current status of PR). I'm probably doing something wrong, but not sure what that is... WDYT?

JoelSpeed replied:
Which gate is `cloudLoadBalancerConfig` behind? You will need to either include or exclude the gate in your test featuregates list.

Most likely, for a TP feature, you want to include it, which will prevent the test from running against the default version of the CRD, where that field is not present (I assume it's still in TP)

patrickdillon replied:
Ok I did that in 0d10a569e66f9a83f032645dd9fdcbd6ec13fbf6 and it seemed to resolve the issue. Thanks!

As this is in the AAA_ungated.yaml test file, it seems like we should potentially skip the non-default feature sets? Not sure if that is possible. A downside of doing that, though, is the tests would need to be updated upon graduation of features to the default feature set... 

I'm definitely fine with this, if you are.

JoelSpeed replied:
In general we want the tests to run against all versions so that we ensure the ungated tests do not break because of a new gated addition

patrickdillon replied:
In this case, that is impossible, right? I suppose if the gate did not have default statuses, then it might not break the tests; but would it be exercising/proving anything?

Just asking for my own understanding/potential process improvement.

JoelSpeed replied:
Yes, for the tests you're adding, because they leverage part of the API, which has a newly gated default value, it makes it impossible.

If you were to split the new tests into a separate file, the remaining ungated tests should in theory be able to continue to run.

I believe that's possible without anything complaining 🤔

patrickdillon reviewed (commented):
Two steps forward, one step back...

Not sure what to do about these test failures for the `cloudLoadBalancerConfig`...

everettraven reviewed (approved):
API changes and test cases LGTM. Tagging in @JoelSpeed for a review.

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -528,18 +528,23 @@ type AWSPlatformStatus struct {
 type AWSResourceTag struct {
-	// key is the key of the tag
+	// key sets the key of the AWS resource tag key-value pair. Key is required when defining an AWS resource tag.
+	// Minimum length of 1. Maximum length of 128.
+	// The string can contain only the set of alphanumeric characters, space (' '), '_', '.', '/', '=', '+', '-', ':', '@' and cannot be prefixed with 'aws:'.
```
For consistency with other APIs and how they are worded, I would suggest
```suggestion
	// Key should consist of between 1 and 128 characters, and may
	// contain only the set of alphanumeric characters, space (' '), '_', '.', '/', '=', '+', '-', ':', and '@'.
	// Key must not be start with 'aws:'.
```

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -528,18 +528,23 @@ type AWSPlatformStatus struct {
 	Key string `json:"key"`
-	// value is the value of the tag.
+	// value sets the value of the AWS resource tag key-value pair. Value is required when defining an AWS resource tag.
+	// The string can contain only the set of Unicode letters, digits, white-space, '_', '.', '/', '=', '+', '-', ':','@'.
```
This wording is different from the above wording, yet the regex is the same, lets use the same wording.
```suggestion
	// Value should consist of between 1 and 256 characters, and may
	// contain only the set of alphanumeric characters, space (' '), '_', '.', '/', '=', '+', '-', ':', and '@'.
```

JoelSpeed reviewed (commented):
Since we are adding the new restriction for not starting with `aws:`, I'd like to see some ratcheting tests added that show that a value persisted that is incompatible with this validation does not negatively affect writes to the rest of the object.

You can see an example of how I'm doing that elsewhere in https://github.com/openshift/api/pull/2142

You'd have to remove the XValidation, persist an invalid value, and then try various updates to check that they still pass (or fail if appropriate)

patrickdillon commented:
> Since we are adding the new restriction for not starting with `aws:`, I'd like to see some ratcheting tests added that show that a value persisted that is incompatible with this validation does not negatively affect writes to the rest of the object.

I have created the test in a7846503fcaa2667c516eb76190f0b2bb674e5bd but the validation does seem to fail on any updates to the object. I tried adding another resourceTag, adding a service endpoint,and finally just persisting the object as is (which is in the commit). All of these fail with 

```
  [38;5;9m[FAILED] unexpected error updating status
  Unexpected error:
      <*errors.StatusError | 0xc0043bd7c0>: 
      Infrastructure.config.openshift.io "test-crmjf" is invalid: status.platformStatus.aws.resourceTags[0].key: Invalid value: "string": the prefix 'aws:' is reserved for AWS system usage and cannot be used at the beginning of a key
```

AFAICT the tests are correct, so it seems like the new validation could be breaking updates?

everettraven commented:
@patrickdillon I had missed that this was a field on the status subresource so you are likely running into a bug that https://github.com/openshift/kubernetes/pull/2167 is supposed to fix

patrickdillon commented:
> @patrickdillon I had missed that this was a field on the status subresource so you are likely running into a bug that [openshift/kubernetes#2167](https://github.com/openshift/kubernetes/pull/2167) is supposed to fix

Ok so I broke the aws prefix validation into a separate PR, https://github.com/openshift/api/pull/2183

everettraven reviewed (approved):
LGTM

JoelSpeed reviewed (commented):
/lgtm

patrickdillon commented:
Sorry I missed that this got tagged and ended up needing to rebase. 

@JoelSpeed if I could trouble you for another look and lgtm

patrickdillon commented:
also it would be fine with me if we want to squash the commits with tide. not sure if there is a preference

patrickdillon commented:
/retest

everettraven commented:
/lgtm

everettraven commented:
Not sure who needs to add the `acknowledge-critical-fixes-only` label. I suppose @JoelSpeed . Is this change something that would fit the scope of a "critical fix"?

patrickdillon commented:
/label acknowledge-critical-fixes-only

patrickdillon commented:
/cherry-pick release-4.18
